src/libraries/System.Linq/src/System/Linq/MaxMin.cs

```diff
@@ -0,0 +1,118 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Numerics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Runtime.Intrinsics;
+
+namespace System.Linq
+{
+    public static partial class Enumerable
+    {
+        private interface IMinMaxCalc<T> where T : struct, IBinaryInteger<T>
+        {
+            public static abstract bool Compare(T left, T right);
+            public static abstract Vector128<T> Compare(Vector128<T> left, Vector128<T> right);
+            public static abstract Vector256<T> Compare(Vector256<T> left, Vector256<T> right);
+        }
+
+        private static T MinMaxInteger<T, TMinMax>(this IEnumerable<T> source)
+            where T : struct, IBinaryInteger<T>
+            where TMinMax : IMinMaxCalc<T>
+        {
+            T value;
+
+            if (source.TryGetSpan(out ReadOnlySpan<T> span))
+            {
+                if (span.IsEmpty)
+                {
+                    ThrowHelper.ThrowNoElementsException();
+                }
+
+                if (!Vector128.IsHardwareAccelerated || span.Length < Vector128<T>.Count)
```

Comment:
it might be dumb question, I'm newbie to vectorized code but does it make sense to do vectorization when you have less than 2 lengths of `Vector128<T>.Count`? it seem it will just load first chunk then find best out of that vector and not actually do any vectorization.

Reply:
It's fine if they overlap.

Reply:
given the answer in the other question wouldn't it make sense to check for i.e. Count + N elements? don't know what the right N is but seems that doing the vectorization for exactly number of elements which fit in vector seem pointless

Reply:
I think you're assuming it's the same cost to iterate through Count values from an array as it is to iterate through Count values from a vector.  It's not.

|    Method |      Mean |     Error |    StdDev | Ratio |
|---------- |----------:|----------:|----------:|------:|
| MaxScalar | 10.022 ns | 0.0474 ns | 0.0420 ns |  1.00 |
| MaxVector |  5.596 ns | 0.0446 ns | 0.0395 ns |  0.56 |

```C#
private static byte[] _bytes = Enumerable.Range(0, Vector128<byte>.Count).Select(i => (byte)i).ToArray();

[Benchmark(Baseline = true)]
public byte MaxScalar()
{
    byte[] bytes = _bytes;
    byte max = bytes[0];
    for (int i = 1; i < bytes.Length; i++)
        if (bytes[i] > max)
            max = bytes[i];
    return max;
}

[Benchmark]
public byte MaxVector()
{
    Vector128<byte> vector = Vector128.LoadUnsafe(ref MemoryMarshal.GetArrayDataReference(_bytes));
    byte max = vector[0];
    for (int i = 1; i < Vector128<byte>.Count; i++)
        if (vector[i] > max)
            max = vector[i];
    return max;
}
```

Reply:
Hm, running this benchmark on my machine shows quite the opposite (`MaxVector = 1.25 * MaxScalar`), and looking at the x64 machine code _MaxScalar_ has no repetitive storing and reading the vector from stack (.NET daily from last week).

<details>
  <summary>machine code</summary>

```assembly
; Bench.MaxScalar()
       sub       rsp,28
       mov       rax,1888D806358
       mov       rax,[rax]
       mov       edx,[rax+8]
       test      edx,edx
       je        short M00_L03
       movzx     ecx,byte ptr [rax+10]
       mov       r8d,1
       cmp       edx,1
       jle       short M00_L02
M00_L00:
       mov       r9d,r8d
       movzx     r9d,byte ptr [rax+r9+10]
       cmp       r9d,ecx
       jle       short M00_L01
       mov       ecx,r9d
M00_L01:
       inc       r8d
       cmp       edx,r8d
       jg        short M00_L00
M00_L02:
       mov       eax,ecx
       add       rsp,28
       ret
M00_L03:
       call      CORINFO_HELP_RNGCHKFAIL
       int       3
; Total bytes of code 77

; Bench.MaxVector()
       sub       rsp,18
       vzeroupper
       xor       eax,eax
       mov       [rsp+8],rax
       mov       [rsp+10],rax
       mov       rax,25C67C06358
       mov       rax,[rax]
       cmp       [rax],al
       vmovdqu   xmm0,xmmword ptr [rax+10]
       vpextrb   eax,xmm0,0
       mov       edx,1
       nop       dword ptr [rax+rax]
       nop       dword ptr [rax+rax]
M00_L00:
       vmovupd   [rsp+8],xmm0                   ; should be hoisted out of loop
       movzx     ecx,byte ptr [rsp+rdx+8]
       cmp       ecx,eax
       jle       short M00_L01
       mov       eax,ecx
M00_L01:
       inc       edx
       cmp       edx,10
       jl        short M00_L00
       add       rsp,18
       ret
; Total bytes of code 93
```
</details>

Based on the numbers I see, then I think @krwq is right with his suggestion similar to
```suggestion
                if (!Vector128.IsHardwareAccelerated || span.Length < 2 * Vector128<T>.Count)
```

But your machine is newer and faster than mine, so maybe that cpu has better store-forwarding logic thus the stack-use here doesn't show up? Or a plain alignment issue for store/read?
I'm just wondering about that huge difference (twice as fast to a quarter slower).

Reply:
> What do you get?

|    Method |     Mean | Ratio |
|---------- |---------:|------:|
| MaxScalar | 15.41 ns |  1.00 |
| MaxVector | 10.64 ns |  0.69 |

---

Forgot that this is the generic variant. For `int` the JIT fully unrolls the loop for the vector-reduction, so this is faster. And I assume that `int` is used more often than `byte`, so leave the code as is (anyway).

Benchmark from https://github.com/dotnet/runtime/pull/76144#discussion_r981303094 with `int`:
|    Method |      Mean | Ratio |
|---------- |----------:|------:|
| MaxScalar | 1.6068 ns |  1.00 |
| MaxVector | 0.5063 ns |  0.32 |